[Clang] Emit error for duplicate mangled names within a lambda

kmclaughlin-arm opened the pull request:
When functions are passed as arguments to a lambda, it's possible for
the mangled names of these functions to be the same despite the prototypes
being different. For example:

```
  int non_streaming_fn(int);
  int streaming_fn(int) __arm_streaming;

  auto lambda_fn = [](const auto &f) { return f(); };
  return lambda_fn(non_streaming_fn) + lambda_fn(streaming_fn);
```

Only one function will be generated for the lambda above and the __arm_streaming
attribute from streaming_fn will be incorrectly applied to the non_streaming_fn call.
With this change, an error will be emitted when a duplicate mangled name is found
but the function prototypes do not match.

kmclaughlin-arm committed:
[Clang] Emit error for duplicate mangled names within a lambda

When functions are passed as arguments to a lambda, it's possible for
the mangled names of these functions to be the same despite the prototypes
being different. For example:

  int non_streaming_fn(int);
  int streaming_fn(int) __arm_streaming;

  auto lambda_fn = [](const auto &f) { return f(); };
  return lambda_fn(non_streaming_fn) + lambda_fn(streaming_fn);

Only one function will be generated for the lambda above and the __arm_streaming
attribute from streaming_fn will be incorrectly applied to non_streaming_fn.
With this change, an error will be emitted when a duplicate mangled name is
found but the function prototypes do not match.

efriedma-quic commented:
Isn't this a bug in the mangler? I mean, it's better to print an error rather than silently miscompile, but this doesn't really solve the issue.

efriedma-quic commented:
> The SME type attributes are not part of the name mangling

If `int(&)() __arm_streaming` is a different type from `int(&)()` for template instantiation, it should have different mangling.  If it doesn't, that's a bug.  If there is no spec for the correct mangling, someone should write a spec.

efriedma-quic commented:
(You shouldn't need to special-case lambdas; any use of the type needs the appropriate mangling, except maybe the function declaration itself.)

sdesmalen-arm commented:
> > The SME type attributes are not part of the name mangling
> 
> If `int(&)() __arm_streaming` is a different type from `int(&)()` for template instantiation, it should have different mangling. If it doesn't, that's a bug. If there is no spec for the correct mangling, someone should write a spec.

I agree that it would be better for the SME attributes to be represented in the type mangling, although changing the mangling at this point would be an ABI break. I'll probe and follow this up!

Nevertheless I believe there is merit to this patch because Clang should not silently generate wrong code. For other instantiations Clang just emits a diagnostic, e.g.

```
void fn(int);
void fs(int) __arm_streaming;

template <typename F, typename... Args>
void dispatcher(F f, Args... args) {
  f(args...);
}

void foo() {
  dispatcher(fn, 41);
  dispatcher(fs, 42);
```

Results in a diagnostic `error: definition with same mangled name '_Z10dispatcherIPFviEJiEEvT_DpT0_' as another definition`. The fact that a similar diagnostic is not produced for lambda instantiations seems like something that needs fixing.

efriedma-quic commented:
Sure, it makes sense to print a diagnostic for lambdas.

-----

I'm having a bit of trouble understanding the way the new code is structured.  What makes the definition of lambda call operators special here?  Do we not call GetOrCreateLLVMFunction with IsForDefinition set?

kmclaughlin-arm committed:
- Rewrite changes to GetOrCreateLLVMFunction

kmclaughlin-arm commented:
> 
> I'm having a bit of trouble understanding the way the new code is structured. What makes the definition of lambda call operators special here? Do we not call GetOrCreateLLVMFunction with IsForDefinition set?

When I added this error I incorrectly thought GetOrCreateLLVMFunction is not called with IsForDefinition set. Now that I've taken another look at it, I see that it is set and the reason the existing error is not emitted is because of the `!Entry->isDeclaration()` condition.

I have removed most of my changes and instead extended the `isDeclaration()` check to also consider whether this is lambda. I don't think it's correct to remove `isDeclaration()`; I believe this was added to allow a non-C++ declaration with a mangled name matching a C++ function (described in https://reviews.llvm.org/D11297).

sdesmalen-arm commented on clang/test/CodeGenCXX/aarch64-sme-lambda-attributes.cpp:
```diff
@@ -0,0 +1,30 @@
+
+#endif
+
+#ifdef TEST2
+
+// expected-error@+3 {{definition with same mangled name '_ZZ36function_params_streaming_compatiblevENK3$_0clIFiiEEEDaRT_' as another definition}}
+// expected-note@+2 {{previous definition is here}}
+int function_params_streaming_compatible() {
+  auto a = [](auto &fn) { return fn(42); };
+  return a(streaming_fn) + a(streaming_compatible_fn);
+}
+
+#endif
```
Can you add a separate test for a non-SME attribute that has a similar issue?

Type attributes for the calling convention are also not part of the type mangling, e.g. https://godbolt.org/z/sEdsGr53K which would (without your change) miscompile.

sdesmalen-arm commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -4652,18 +4652,35 @@ llvm::Constant *CodeGenModule::GetOrCreateLLVMFunction(
+              cast_or_null<FunctionDecl>(OtherGD.getDecl());
+          if (FD && FD->hasPrototype() && OtherFD && OtherFD->hasPrototype()) {
+            if (FD->getType()->getAs<FunctionProtoType>() !=
+                OtherFD->getType()->getAs<FunctionProtoType>())
```
(I see you implemented the same suggestion, not sure why Github didn't publish my comment)

kmclaughlin-arm committed:
- Removed second SME attributes test
- Added a test using calling convention attributes

kmclaughlin-arm committed:
- Renamed aarch64-sme-lambda-attributes.cpp to lambda-instantiation-mangling-conflicts.cpp
- Replaced ifdefs in test with use of split-file

kmclaughlin-arm commented on clang/test/CodeGenCXX/aarch64-sme-lambda-attributes.cpp:
```diff
@@ -0,0 +1,31 @@
+// REQUIRES: aarch64-registered-target
+
+// RUN: %clang_cc1 -triple aarch64 -emit-llvm -o - %s -verify -DTEST1
+// RUN: %clang_cc1 -triple aarch64 -target-feature +sme -target-feature +sme2 -emit-llvm -o - %s -verify -DTEST2
```
I've removed the ifdefs & renamed this to `lambda-instantiation-mangling-conflicts.cpp` as it was a bit shorter, but I can rename to `lambda-instantiation-require-different-mangled-names.cpp` if preferred.

kmclaughlin-arm commented:
> For non-lambda methods, the way this works it that we call GetOrCreateLLVMFunction for both methods... for the first method, the `!Entry->isDeclaration()` would be false, but for the second one, it would be true because we've emitted the definition of the first method. What's different for lambda methods?

I don't think lambda methods are any different to what you have described, but I think my previous explanation was unclear. What I found is that IsForDefinition is set only the second time GetOrCreateLLVMFunction is called, where `Entry->isDeclaration()` is also true.

efriedma-quic commented:
In the example in https://github.com/llvm/llvm-project/pull/107581#issuecomment-2343651051 , there are four relevant calls to GetOrCreateLLVMFunction: one for each function with IsInDefinition false, and one for each function with IsInDefinition true.  The last of those calls triggers the diagnostic

For the lambda example, there are only three relevant calls to GetOrCreateLLVMFunction; one for each function with IsInDefinition false, but then only one with IsInDefinition true.

It's not clear to me why the two cases are different, and I don't really want to add lambda-specific check without knowing why lambdas are special here.  Maybe the right condition to check is something else.

kmclaughlin-arm commented:
> For the lambda example, there are only three relevant calls to GetOrCreateLLVMFunction; one for each function with IsInDefinition false, but then only one with IsInDefinition true.
> 
> It's not clear to me why the two cases are different, and I don't really want to add lambda-specific check without knowing why lambdas are special here. Maybe the right condition to check is something else.

From what I've observed comparing the lambda and template test cases, the difference comes from how we create the list of Decls which can be deferred. 

Both functions `foo()` from the template case in https://github.com/llvm/llvm-project/pull/107581#issuecomment-2343651051 and `function_params_normal_streaming()` from the lambda test case in this PR cannot be deferred by EmitGlobal here and must be emitted straight away (by calling GetOrCreateLLVMFunction):
```
 if (MustBeEmitted(Global) && MayBeEmittedEagerly(Global)) {
    EmitGlobalDefinition(GD);
    addEmittedDeferredDecl(GD);
    return;
```
For the template case this happens first, before calling EmitGlobal for `dispatcher()`. For lambdas the function arguments of the lambda are handled first. This changes whether the name is found later in EmitGlobal when adding the MangledName to the DeferredDeclsToEmit list:
```
  StringRef MangledName = getMangledName(GD);
    if (GetGlobalValue(MangledName) != nullptr) {
      // The value has already been used and should therefore be emitted.
      addDeferredDeclToEmit(GD);
    ...
    } else {
      // Otherwise, remember that we saw a deferred decl with this name.  The
      // first use of the mangled name will cause it to move into
      // DeferredDeclsToEmit.
      DeferredDecls[MangledName] = GD;
    }
```
Although we'll set DeferredDecls with the mangled name from the lambda twice, it will only appear in DeferredDeclsToEmit once. In the other example, it will be added to the DeferredDeclsToEmit list with addDeferredDeclToEmit both times. When EmitDeferred is called later it uses this list and calls GetOrCreateLLVMFunction with IsForDefinition set to true.

I'm not sure adding a check specificially for lambdas is the right thing to do because of this. I considered trying to emit the error somewhere in lib/Sema, however I think this is too early to check whether functions will have the same mangled name.

Something I'm considering instead is to try and emit the error in EmitGlobal, where we set DeferredDecls for a given MangledName (around `DeferredDecls[MangledName] = GD` in the snippet above). I think this could be a better place to emit the error as we can look for whether GD has already been assigned to MangledName. I'm still investigating how to do this without causing failures in some of the existing tests which are using lambdas & templates.

I will also be away for the next two weeks but plan to pick this up again when I return.

efriedma-quic commented:
Oh, that makes sense... so the issue is generally with functions we emit lazily?

I'd say it's reasonable to emit an error if we have two definitions for the same symbol, even if we don't end up emitting them because they're deferred.  Lazy emission is an optimization, and we don't guarantee whether a particular odr-used function will be emitted.

kmclaughlin-arm committed:
- Moved error to EmitGlobal and removed specific check for a lambda

kmclaughlin-arm commented:
> Oh, that makes sense... so the issue is generally with functions we emit lazily?
> I'd say it's reasonable to emit an error if we have two definitions for the same symbol, even if we don't end up emitting them because they're deferred.

Yes, I believe so. I've moved the error to EmitGlobal at the point where DeferredDecls is updated if a duplicate mangled name is found. I've tried to exclude the cases where this can be expected (such as for multiversioning where the mangled name may be changed later by UpdateMultiVersionNames), but there is nothing specific to lambdas now.

> I agree that it would be better for the SME attributes to be represented in the type mangling, although changing the mangling at this point would be an ABI break. I'll probe and follow this up!

Following on from this, I am also working on changes to name mangling of types which have SME attributes (https://github.com/ARM-software/abi-aa/pull/290)

efriedma-quic commented:
Can you add a testcase that doesn't involve lambdas?  Maybe something like the following:

```
inline void f(int) asm("foo");
inline void f(int) {}
inline void f() asm("foo");
inline void f(){}
```

kmclaughlin-arm committed:
- Add test without lambdas to duplicate-mangled-name.cpp

efriedma-quic reviewed (approved):
LGTM
